modules/FvwmCascade/FvwmCascade.man
.\" t
.\" @(#)FvwmCascade.1  11/27/96
.de EX		\"Begin example
.ne 5
.if n .sp 1
.if t .sp .5
.nf
.in +.5i
..
.de EE
.fi
.in -.5i
.if n .sp 1
.if t .sp .5
..
.TH FvwmCascade 1 "November 27, 1996" "FvwmCascade 1.0" "FvwmCascade 1.0"
.UC
.SH NAME
FvwmCascade \- layer FVWM windows
.SH SYNOPSIS
FvwmCascade is spawned by fvwm, so no command line invocation will work.

.SH DESCRIPTION
This module attempts to cascade windows on the current screen
subject to certain constraints.  Layering is performed so consecutive
windows will have their window titles visible underneath the previous.

.SH INVOCATION
FvwmCascade is best invoked from a menu, popup or button.  There are a
number of command line options which can be used to constrain the
layering.  These are described below.  As an example case, one could
call FvwmCascade with the following arguments:
.EX
FvwmCascade \-resize 10 2 80 70
.EE

This invocation will cascade windows starting 10 by 2 percent into and
down the screen.  Windows will be constrained to 80 by 70 percent of
the screen dimensions.  Since the \fIresize\fP is also specified,
windows will be resized to the given constrained width and height.

Command-line arguments passed to FvwmCascade are described here.
.IP \-a
Causes \fIall\fP window styles to be affected, even ones with the
WindowListSkip style.
.IP \-desk
Causes all windows on the desk to be cascaded instead of the current
screen only.
.IP \-flatx
Inhibits border width increment.
.IP \-flaty
Inhibits border height increment.
.IP "\-incx \fIarg\fP"
Specifies a horizontal increment which is successively added to
cascaded windows.  \fIarg\fP is a percentage of screen width, or pixel
value if a \fIp\fP is suffixed.  Default is zero.
.IP "\-incy \fIarg\fP"
Specifies a vertical increment which is successively added to cascaded
windows.  \fIarg\fP is a percentage of screen height, or pixel value
if a \fIp\fP is suffixed.  Default is zero.
.IP \-m
Causes maximized windows to also be affected (implied by \-all).
.IP \-noraise
Inhibits window raising, leaving the depth ordering intact.
.IP \-nostretch
Inhibits window expansion when using the \-resize option.  Windows
will only shrink to fit the maximal width and height (if given).
.IP \-r
Reverses the window sequence.
.IP \-s
Causes sticky windows to also be affected (implied by \-all).
.IP \-resize
Forces all windows to resize to the constrained width and height (if
given).
.IP \-t
Causes transient windows to also be affected (implied by \-all).
.IP \-u
Causes untitled windows to also be affected (implied by \-all).

Up to four numbers can be placed on the command line that are not
switches.  The first pair specify an x and y offset to start the first
window (default is 0, 0).  The second pair specify a maximal width and
height for the layered windows.  If an affected window exceeds either
this width or height, it is resized to the maximal width or height.
If each number is suffixed with the letter p, then it is taken to be a
pixel value, otherwise it is interpreted as a screen percentage.
Specifying zero for any parameter is equivalent to not specifying it.

.SH BUGS
It is probably not a good idea to delete windows while windows are
being cascaded.

.SH AUTHOR
Andrew Veliath


NEWS
Note, the changes for the last STABLE release start with release
2.6.0.

-------------------------------------------------------------------
Changes in stable release 2.6.8 (UNRELEASED)

* Bug fixes:

  - Various DESTDIR fixes (especially around the default-config
    Makefile)
  - fvwm-perllib pod2man fixes.
  - FvwmIconMan no longer triggers a warning about bad size hints
    in the fvwm core.
  - VMware windows do not disapper when switching between
    fullscreen and normal state.
  - Fix (de)installation with the configure options
    --program-prefix, --program-suffix and
    --program-transform-name.

-------------------------------------------------------------------
Changes in stable release 2.6.7 (06-Mar-2016)

* New fvwm features:

  - A new default configuration which is available when fvwm
    doesn't detect a configuration file to load.
  - A new conitional command "Desk n" can restrict matching
    windows to a specific desk.
  - A new conditional command "Screen n" to restrict matching
    windows on a given Xinerama screen.
  - New expansion variable "w.screen" to ascertain the Xineram
    screen number a window is on.
  - New command "InfoStoreClear" to remove all items in the
    InfoStore.

* Removed features:

  - The old and unmaintained debian/and rpm/ directories have
    been remmoved; use the maintainers' copies where available.
  - VMS support has been removed.
  - GTK1.x support has been removed.
  - GNOME-specific window hints (pre-EWMH) have been removed.
  - Some fvwm modules have been removed:
      - FvwmDragWell   (no replacement)
      - FvwmGTK        (no replacement)
      - FvwmSave       (no replacement)
      - FvwmSaveDesk   (no replacement)
      - FvwmScroll     (no replacement)
      - FvwmTabs       (no replacement, never worked anyway)
      - FvwmTaskBar    (use FvwmButtons)
      - FvwmTheme      (in core of fvwm as colorsets)
      - FvwmWharf      (use FvwmButtons)
      - FvwmWinList    (use WindowList command)
      - FvwmWindowMenu (use WindowList command)
      - FvwmIconBox    (use the IconBox style instead)

* New module features:

  - FvwmButtons learned a new option "Colorset" to its
    ChangeButton command.

* fvwm-menu-desktop updated:

  - Renamed default menu to XDGMenu and changed the name
    of the FvwmForm to FvwmForm-XDGMenu-Config to not conflict
    with someone already using FvwmMenu.
  - fvwm-menu-desktop will now load defaults from the
    FvwmForm-XDGMenu-Config data file.
  - Improved dynamic menus by regenerating them on-the-fly.
  - Added new options: --regen-cmd, --dynamic, and more.

* Bug fixes:

  - A bug introduced in 2.6.6 could cause applications with
    negative coordinates to be placed at strange positions.  This
    affected for example acroread when switching to fullscreen
    mode.  This has been fixed.
  - FvwmButtons "Silent" option for dynamic ChangeButton commands
    no longer loops infinitely.

Changes in stable release 2.6.6 (15-Mar-2016)

* New fvwm features:

  - Support for Russian from Ivan Gayevskiy.
  - EnvMatch supports infostore variables.
  - The option "forget" to the Maximize command allows to
    unmaximize a window without changing its size and position.
  - Windows shaded to a corner (NW, NE, SW, SE) are reduced to a
    small square.
  - New option "!raise" to the WarpToWindow command.
  - The new extended variables $[wa.x], $[wa.y], $[wa.width],
    $[wa.height] can be used to get the geometry of the EWMH
    working area, and $[dwa.x], $[dwa.y], $[dwa.width],
    $[dwa.height] can be used to get the geometry of the EWMH
    dynamic working area.
  - The Resize commands accept "wa" or "da" as a suffix of the
    width or height value.  If present, the value is a percentage
    of the width or height of the EWMH working area or the EWMH
    dynamic working area.
  - Fvwm is much more resilient against applications that flood the
    window manager with repeated events.

* New module features:

  - FvwmForm supports separator lines.
  - New FvwmIconMan options:
      IconAndSelectButton
      IconAndSelectColorset

* Bug fixes:

  - Provide a wrapper for the deprecation of XKeycodeToKeysym and
    use XkbKeycodeToKeysym() where appropriate.
  - fvwm-menu-desktop is re-written and provides better support of
    the XDG menu specification.
  - Fix fvwm-menu-desktop keyError bug.  Use "others" if no desktop
    environment found.
  - FvwmIconMan had problems displaying the hilight colour on some
    systems (64 bit issue?).
  - Globally active windows cannot take the focus if the style
    forbids programs to take focus themselves (style
    !FPFocusByProgram).
  - Windows no longer jump from one position to the other which
    could happen in some cases with SnapAttraction.  Windows now
    snap to the closest window (or screen edge).
  - Removing bindings had several strange side effects that are
    fixed now (removing too many bindings; old bindings showing up
    again after another is removed; possibly other effects).
  - Windows sometimes did not get expose events (i.e. did not
    redraw properly) if they were uncovered by moving a window
    above them.  This has been fixed.
  - FvwmConsole now causes much less network traffic.
  - Suppress bogus events sent to the modules when a window is
    resized with the mouse.
  - Properly handle the has_ref_window_moved flag for ResizeMove and
    ResizeMoveMaximize.
  - Removed some unnecessary redraws in FvwmPager.
  - The option "NoDeskLimitY" option of the GotoPage command did
    not work.
  - Negative coordinates in the "rectangle" option to the Menu
    commend did not work correctly.  This has been fixed.
  - Removes a slight graphics problem whith the ResizeMaximize
    command being invoked from a window button menu.
  - When an attempt to reparent a client window (i.e. decorate
    it) fails, fvwm no longer throws away all events but only the
    events for that window.

modules/FvwmCascade/FvwmCascade.c
/*************************************************************************
 * FvwmCascade.c -- fvwm module to cascade windows
 *
 * See man page for usage.
 *
 * Written by Andrew Veliath
 * Copyright 1996
 * 
 * $Id$
 ************************************************************************/
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>
#include <sys/time.h>
#include <sys/types.h>
#include <unistd.h>
#include <signal.h>

#ifdef ISC
#include <sys/bsdtypes.h>
#endif

#if defined ___AIX || defined _AIX || defined __QNX__ || defined ___AIXV3 || defined AIXV3 || defined _SEQUENT_
#include <sys/select.h>
#endif

#include <X11/Xlib.h>

#include "../../libs/fvwmlib.h"
#include "../../fvwm/module.h"
#include "../../fvwm/fvwm.h"

typedef struct window_item {
    Window frame;
    int th, bw;
    unsigned long width, height;
    struct window_item *prev, *next;
} window_item, *window_list;

/* vars */
Display *dpy;
int dwidth, dheight;
char *argv0;
int fd[2], fd_width;
window_list wins = NULL, wins_tail = NULL;
int wins_count = 0;
FILE *console;

/* switches */
int ofsx = 0, ofsy = 0;
int maxw = 0, maxh = 0;
int untitled = 0, transients = 0;
int maximized = 0;
int all = 0;
int desk = 0;
int reversed = 0, raise_window = 1;
int resize = 0, nostretch = 0, sticky = 0;
int flatx = 0, flaty = 0;
int incx = 0, incy = 0;

void insert_window_list(window_list *wl, window_item *i)
{
    if (*wl) {
	if ((i->prev = (*wl)->prev))
	    i->prev->next = i;
	i->next = *wl;
	(*wl)->prev = i;
    } else
	i->next = i->prev = NULL;
    *wl = i;
}

void free_window_list(window_list *wl)
{
    window_item *q;
    while (*wl) {
	q = *wl;
	free(*wl);
	*wl = (*wl)->next;
    }
}

int is_suitable_window(unsigned long *body)
{
    XWindowAttributes xwa;
    unsigned long flags = body[8];

    if ((flags&WINDOWLISTSKIP) && !all)
	return 0;

    if ((flags&MAXIMIZED) && !maximized)
	return 0;
    
    if ((flags&STICKY) && !sticky)
	return 0;
    
    if (!XGetWindowAttributes(dpy, (Window)body[1], &xwa))
	return 0;
    
    if (xwa.map_state != IsViewable)
	return 0;

    if (!(flags&MAPPED))
	return 0;

    if (flags&ICONIFIED)
	return 0;

    if (!desk) {
	int x = (int)body[3], y = (int)body[4];
	int w = (int)body[5], h = (int)body[6];
	if (!((x < dwidth) && (y < dheight)
	      && (x + w > 0) && (y + h > 0)))
	    return 0;
    }
	    
    if (!(flags&TITLE) && !untitled)
	return 0;

    if ((flags&TRANSIENT) && !transients)
	return 0;

    return 1;
}

int get_window(void)
{
    unsigned long header[HEADER_SIZE], *body;
    int count, last = 0;
    fd_set infds;
    FD_ZERO(&infds);
    FD_SET(fd[1], &infds);
    select(fd_width,&infds, 0, 0, NULL);
    if ((count = ReadFvwmPacket(fd[1],header,&body)) > 0) {
	switch (header[1])
	{
	case M_CONFIGURE_WINDOW:
	    if (is_suitable_window(body)) {
		window_item *wi =
		    (window_item*)safemalloc(sizeof( window_item ));
		wi->frame = (Window)body[1];
		wi->th = (int)body[9];
		wi->bw = (int)body[10];
		wi->width = body[5];
		wi->height = body[6];
		if (!wins_tail) wins_tail = wi;
		insert_window_list(&wins, wi);
		++wins_count;
	    }
	    last = 1;
	    break;
	    
	case M_END_WINDOWLIST:
	    break;
	    
	default:
	    fprintf(console, 
		    "%s: internal inconsistency: unexpected message\n",
		    argv0);
	    break;
	}
	free(body);
    }
    return last;
}

void wait_configure(window_item *wi)
{
    int found = 0;
    unsigned long header[HEADER_SIZE], *body;
    int count;
    fd_set infds;
    FD_ZERO(&infds);
    FD_SET(fd[1], &infds);
    select(fd_width,&infds, 0, 0, NULL);
    while (!found)	
	if ((count = ReadFvwmPacket(fd[1],header,&body)) > 0) {
	    if  ((header[1] == M_CONFIGURE_WINDOW)
		 && (Window)body[1] == wi->frame)
		found = 1;
	    free(body);
	}
}

int atopixel(char *s, unsigned long f)
{
    int l = strlen(s);
    if (l < 1) return 0;
    if (isalpha(s[l - 1])) {
	char s2[24];
	strcpy(s2,s);
	s2[strlen(s2) - 1] = 0;	
	return atoi(s2);
    }
    return (atoi(s) * f) / 100;
}

void cascade_windows(void)
{
    char msg[128];
    int cur_x = ofsx, cur_y = ofsy;
    window_item *w = reversed ? wins_tail : wins;
    while (w)
    {
	unsigned long nw = 0, nh = 0;
	if (raise_window)
	    SendInfo(fd,"Raise",w->frame);
	sprintf(msg, "Move %up %up", cur_x, cur_y);
	SendInfo(fd,msg,w->frame);
	if (resize) {
	    if (nostretch) {		
		if (maxw
		    && (w->width > maxw))
		    nw = maxw;
		if (maxh
		    && (w->height > maxh))
		    nh = maxh;
	    } else {
		nw = maxw; 
		nh = maxh;
	    }
	    if (nw || nh) {
		sprintf(msg, "Resize %lup %lup", 
			nw ? nw : w->width, 
			nh ? nh : w->height);
		SendInfo(fd,msg,w->frame);
	    }
	}
	wait_configure(w);
	if (!flatx)
	    cur_x += w->bw;
	cur_x += incx;
	if (!flaty)
	    cur_y += w->bw + w->th;
	cur_y += incy;
	w = reversed ? w->prev : w->next;
    }
}

void parse_args(char *s, int argc, char *argv[], int argi)
{
    int nsargc = 0;
    /* parse args */
    for (; argi < argc; ++argi)
    {
	if (!strcmp(argv[argi],"-u")) {
	    untitled = 1;
	}
	else if (!strcmp(argv[argi],"-t")) {
	    transients = 1;
	}
	else if (!strcmp(argv[argi], "-a")) {
	    all = untitled = sticky = transients = maximized = 1;
	}
	else if (!strcmp(argv[argi], "-r")) {
	    reversed = 1;
	}
	else if (!strcmp(argv[argi], "-noraise")) {
	    raise_window = 0;
	}
	else if (!strcmp(argv[argi], "-desk")) {
	    desk = 1;
	}
	else if (!strcmp(argv[argi], "-flatx")) {
	    flatx = 1;
	}
	else if (!strcmp(argv[argi], "-flaty")) {
	    flaty = 1;
	}
	else if (!strcmp(argv[argi], "-m")) {
	    maximized = 1;
	}
	else if (!strcmp(argv[argi], "-s")) {
	    sticky = 1;
	}
	else if (!strcmp(argv[argi], "-resize")) {
	    resize = 1;
	}
	else if (!strcmp(argv[argi], "-nostretch")) {
	    nostretch = 1;
	}
	else if (!strcmp(argv[argi], "-incx") && ((argi + 1) < argc)) {
	    incx = atopixel(argv[++argi], dwidth);
	}
	else if (!strcmp(argv[argi], "-incy") && ((argi + 1) < argc)) {
	    incy = atopixel(argv[++argi], dheight);
	}
	else {
	    if (++nsargc > 4) {
		fprintf(console,
			"%s: %s: ignoring unknown arg %s\n",
			argv0, s, argv[argi]);
		continue;
	    }
	    if (nsargc == 1) {
		ofsx = atopixel(argv[argi], dwidth);
	    } else if (nsargc == 2) {
		ofsy = atopixel(argv[argi], dheight);
	    } else if (nsargc == 3) {
		maxw = atopixel(argv[argi], dwidth);
	    } else if (nsargc == 4) {
		maxh = atopixel(argv[argi], dheight);
	    }	    
	}
    }
}

#ifdef USERC
int parse_line(char *s, char ***args)
{
    int count = 0, i = 0;
    char *arg_save[48];
    strtok(s, " ");
    while ((s = strtok(NULL, " ")))
	arg_save[count++] = s;
    *args = (char **)safemalloc(sizeof( char * ) * count);
    for (; i < count; ++i)
	(*args)[i] = arg_save[i];
    return count;
}

#ifdef FVWM1
char *GetConfigLine(char *filename, char *match)
{
    FILE *f = fopen(filename, "r");
    if (f) {
	int l = strlen(match), found = 0;
	char line[256], *s = line;
	line[0] = 0;
	s = fgets(line, 256, f);
	while (s && !found) {
	    if (strncmp(line, match, l) == 0) {
		found = 1;
		break;
	    }
	    s = fgets(line, 256, f);
	}
	fclose(f);
	if (found) {
	    char *ret;
	    int l2 = strlen(line);
	    ret = (char *)safemalloc(sizeof(char) * l2);
	    strcpy(ret, line);
	    if (ret[l2 - 1] == '\n')
		ret[l2 - 1] = 0;
	    return ret;
	} else
	    return NULL;
    } else
	return NULL;
}
#endif /* FVWM1 */
#endif /* USERC */

void DeadPipe(int sig) { exit(0); }

void main(int argc, char *argv[])
{
#ifdef USERC
    char match[128];
    int config_line_count, len;
    char *config_line;
#endif

    console = fopen("/dev/console","w");
    if (!console) console = stderr;

    if (!(argv0 = strrchr(argv[0],'/')))
	argv0 = argv[0];
    else 
	++argv0;

    if (argc < 6) {
	fprintf(stderr,
#ifdef FVWM1
		"%s: module should be executed by fvwm only\n",
#else
		"%s: module should be executed by fvwm2 only\n",
#endif
		argv0);
	exit(-1);
    }

    fd[0] = atoi(argv[1]);
    fd[1] = atoi(argv[2]);

    if (!(dpy = XOpenDisplay(NULL))) {
	fprintf(console, "%s: couldn't open display %s\n",
		argv0,
		XDisplayName(NULL));
	exit(-1);
    }
    signal (SIGPIPE, DeadPipe);

    {
	int s = DefaultScreen(dpy);
	dwidth = DisplayWidth(dpy, s);
	dheight = DisplayHeight(dpy, s);
    }

    fd_width = GetFdWidth();
    
#ifdef USERC
    strcpy(match, "*");
    strcat(match, argv0);
    len = strlen(match);
#ifdef FVWM1
    if ((config_line = GetConfigLine(argv[3], match))) {
	char **args = NULL;
	config_line_count = parse_line(config_line, &args);
	parse_args("config args", 
		   config_line_count, args, 0);
	free(config_line);
	free(args);
    }
#else
    GetConfigLine(fd, &config_line);
    while (config_line != NULL) {
	if (strncmp(match,config_line,len)==0) {
	    char **args = NULL;
	    int cllen = strlen(config_line);
	    if (config_line[cllen - 1] == '\n')
		config_line[cllen - 1] = 0;
	    config_line_count = parse_line(config_line, &args);
	    parse_args("config args", 
		       config_line_count, args, 0);
	    free(args);
	}
	GetConfigLine(fd, &config_line);
    }
#endif /* FVWM1 */
#endif /* USERC */

    parse_args("module args", argc, argv, 6);

#ifdef FVWM1
    {
	char msg[256];
	sprintf(msg, "SET_MASK %lu\n",(unsigned long)(
	    M_CONFIGURE_WINDOW|
	    M_END_WINDOWLIST
	    ));
	SendInfo(fd,msg,0);
	
#ifdef FVWM1_MOVENULL
	/* avoid interactive placement in fvwm version 1 */
	if (!ofsx) ++ofsx;
	if (!ofsy) ++ofsy;
#endif
    }
#else
    SetMessageMask(fd,
		   M_CONFIGURE_WINDOW 
		   | M_END_WINDOWLIST
	);
#endif

    SendInfo(fd,"Send_WindowList",0);
    while (get_window());
    if (wins_count)
	cascade_windows();
    free_window_list(&wins);
    if (console != stderr)
	fclose(console);
}
